dist: tools: add coccinelle scripts

kaspar030 opened the pull request:
This PR adds initial support for testing the codebase against [coccinelle](http://coccinelle.lip6.fr/) semantic patches.

Coccinelle allows checking the codebase for semantic problems, e.g., a forgot mutex unlock in an error case.

Please see the coccinelle homepage for more details.

I currently envision the following concept:

- as with cppcheck, every added or modified file will be checked
- there are two sets of coccinelle patches: "force" and "warn"
- if coccinelle has a hit in "warn", the resulting patch will be printed, but doesn't cause the CI to fail
- if coccinelle has a hit in "force", the CI build fails
- we add semantic patches first to "warn", fix our codebase, and if the patch is confident enough, move it to "force"

Local testing of the whole codebase can be done by executing ```dist/tools/coccinelle/check.sh```, if coccinelle is installed.

~~CI will currently fail as coccinelle is not yet installed in RIOT's docker container (see https://github.com/RIOT-OS/riotdocker/pull/29).~~ (https://github.com/RIOT-OS/riotdocker/pull/29 has been merged)

Comments welcome!

~~(Waiting for #6940)~~

haukepetersen commented:
awesome and +1 for the concept you propose here!

miri64 commented:
#6940 was merged, please rebase.

kaspar030 commented:
- rebased and squashed

miri64 commented on dist/tools/coccinelle/include/riot-standard.h:
```diff
@@ -0,0 +1 @@
+#define LL_SEARCH_SCALAR(pkt, tcp, type, GNRC_NETTYPE_TCP) tcp = XXX
```
What does this do?

kaspar030 replied:
Coccinelle by default ignores macros. In this case, there was a false positive for one of the coccinelle scripts, as coccinelle didn't know that ```LL_SEARCH_SCALAR``` might change ```tcp```.

kaspar030 commented:
Double green!

kYc0o commented:
I'd ACK but I think Martine still has comments.

kaspar030 commented:
Stop, there seems to be a problem with an option missing in the coccinelle version used by the CI. (don't merge yet)

miri64 commented:
> […]  but I think Martine still has comments.

Please remind me what comments I should have?

kYc0o commented:
I was just talking about the question you asked, which I don't know if the answer is OK for you.

kaspar030 commented:
> Stop, there seems to be a problem with an option missing in the coccinelle version used by the CI. (don't merge yet)

I fixed the issue (newer coccinelle allows passing a file list fie --file-groups, CI version does not. So now we pass ```${FILES}``` directly on the commandline, as done for other check scripts.).

- rebased, squashed

IMO this is ready to go.

miri64 commented:
> I was just talking about the question you asked, which I don't know if the answer is OK for you.

Yes. Question was answered, so I'm alright with it. ;-)
